src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -34,41 +36,169 @@ public static class TagHelperDescriptorFactory
 
-        private static IEnumerable<string> GetTagNames(Type tagHelperType)
+        /// <summary>
+        /// Internal for testing.
+        /// </summary>
+        internal static IEnumerable<string> GetCommaSeparatedValues(string text)
         {
-            var typeInfo = tagHelperType.GetTypeInfo();
-            var attributes = typeInfo.GetCustomAttributes<HtmlElementNameAttribute>(inherit: false);
+            // We don't want to remove empty entries, need to notify users of invalid values.
+            return text?.Split(',').Select(tagName => tagName.Trim()) ?? Enumerable.Empty<string>();
+        }
 
-            // If there isn't an attribute specifying the tag name derive it from the name
-            if (!attributes.Any())
-            {
-                var name = typeInfo.Name;
+        /// <summary>
+        /// Internal for testing.
+        /// </summary>
+        internal static bool ValidTargetElementAttributeNames(
+            TargetElementAttribute attribute,
+            ParserErrorSink errorSink)
+        {
+            return
+                ValidateNames(
+                    names: GetCommaSeparatedValues(attribute.Tags),
+                    targetingAttributes: false,
+                    errorSink: errorSink) &&
+                ValidateNames(
+                    names: GetCommaSeparatedValues(attribute.Attributes),
+                    targetingAttributes: true,
+                    errorSink: errorSink);
 
-                if (name.EndsWith(TagHelperNameEnding, StringComparison.OrdinalIgnoreCase))
+        }
+
+        private static bool ValidateNames(
+            IEnumerable<string> names,
+            bool targetingAttributes,
+            ParserErrorSink errorSink)
+        {
+            var targetName = targetingAttributes ?
+                Resources.TagHelperDescriptorFactory_Attribute :
+                Resources.TagHelperDescriptorFactory_Tag;
+            var validNames = true;
+
+            foreach (var name in names)
+            {
+                if (string.IsNullOrWhiteSpace(name))
                 {
-                    name = name.Substring(0, name.Length - TagHelperNameEnding.Length);
+                    errorSink.OnError(
+                        SourceLocation.Zero,
+                        Resources.FormatTargetElementAttribute_NameCannotBeNullOrWhitespace(targetName));
+
+                    validNames = false;
+
+                    continue;
                 }
 
-                return new[] { ToHtmlCase(name) };
+                foreach (var character in name)
+                {
+                    if (char.IsWhiteSpace(character) ||
+                        InvalidNonWhitespaceNameCharacters.Contains(character))
+                    {
+                        errorSink.OnError(
+                            SourceLocation.Zero,
+                            Resources.FormatTargetElementAttribute_InvalidName(
+                                targetName.ToLowerInvariant(),
```

Comment:
`targetName` is locale-sensitive and must be lower-cased using the current culture